fix(core): export inject() from @angular/core

alxhub opened the pull request:
inject() supports the ngInjectableDef-based configuration of the injector
(otherwise known as tree-shakeable services). It was missing from the
public API of @angular/core, this PR adds it.

The test added here is correct in theory, but may pass accidentally due
to the decorator side-effect replacing the inject() call at runtime. An
upcoming compiler PR will strip reified decorators from output entirely.

Fixes #22388.

CC @jelbourn

mary-poppins commented:
You can preview a25dc34 at https://pr22389-a25dc34.ngbuilds.io/.

IgorMinar commented on tools/public_api_guard/core/core.d.ts:
```diff
@@ -447,6 +447,9 @@ export interface HostDecorator {
 export declare const HostListener: HostListenerDecorator;
 
+/** @experimental */
+export declare function inject<T>(token: Type<T> | InjectionToken<T>, notFoundValue?: undefined, flags?: InjectFlags): T;
```
shouldn't this be exported as `ɵinject` instead? I don't think that this needs to be a public api since only generated code needs access to it as far as I can tell. 

Can you please clarify?

alxhub replied:
This is public API. `InjectionToken` depends on it being so. A user would write:

```
const TOKEN = new InjectionToken<Service>('tree-shakeable token!', {
  scope: Module,
  factory: () => new Service(inject(Dep))
});
```

`inject()` is how they interact with the injection system.

IgorMinar replied:
so we don't support providing dependencies via `deps` just like we do any any other provider?

alxhub replied:
Not with `InjectionToken`, because this isn't actually a provider definition. We can't support a provider here because that would require reflecting dependencies, which would bring in `ReflectionCapabilities`.

IgorMinar commented on packages/core/src/di/injector.ts:
```diff
@@ -410,6 +410,16 @@ export function setCurrentInjector(injector: Injector | null): Injector|null {
+/**
+ * Injects a token from the currently active injector.
+ *
+ * This function must be used in the context of an `ngInjectableDef` factory function, and will
```
ngInjectableDef is not a public api so this comment won't make sense to a application developer reading this comment.

alxhub replied:
True, I'll reword it for now.

mary-poppins commented:
You can preview de32278 at https://pr22389-de32278.ngbuilds.io/.

IgorMinar commented:
I spoke with Chuck and we don't think that reflection capabilities would be
needed if we used deps. Can you please sync up with him and come to an
agreement?

My preference is to keep this api private (for now)  unless there is a
strong reason to make it public.

On Thu, Feb 22, 2018, 4:34 PM Mary Poppins <notifications@github.com> wrote:

> You can preview de32278
> <https://github.com/angular/angular/commit/de32278133d318851bc80f54a12e76414b2f7a89>
> at https://pr22389-de32278.ngbuilds.io/.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/22389#issuecomment-367870574>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANM6Ln2M2RttuDqam_rPXVaMtCTrkM3ks5tXgeCgaJpZM4SQKns>
> .
>

alxhub committed:
fix(core): export inject() from @angular/core

inject() supports the ngInjectableDef-based configuration of the injector
(otherwise known as tree-shakeable services). It was missing from the
exported API of @angular/core, this PR adds it.

The test added here is correct in theory, but may pass accidentally due
to the decorator side-effect replacing the inject() call at runtime. An
upcoming compiler PR will strip reified decorators from the output
entirely.

Fixes #22388

mary-poppins commented:
You can preview 50a9b9c at https://pr22389-50a9b9c.ngbuilds.io/.

IgorMinar reviewed (dismissed):
thanks

mary-poppins commented:
You can preview abc6001 at https://pr22389-abc6001.ngbuilds.io/.

mary-poppins commented:
You can preview ef54180 at https://pr22389-ef54180.ngbuilds.io/.

alxhub commented:
After discussing with @mhevery, we're agreed that `inject()` should be exported publicly. It's intended to replace `deps: [...]` as the way of specify dependencies to factory functions, and has the major advantage that its call sites are monomorphic.

mary-poppins commented:
You can preview c86fb35 at https://pr22389-c86fb35.ngbuilds.io/.

IgorMinar commented on packages/core/src/di/injector.ts:
```diff
@@ -410,6 +410,16 @@ export function setCurrentInjector(injector: Injector | null): Injector|null {
+ * Injects a token from the currently active injector.
+ *
+ * This function must be used in the context of a factory function such as one defined for an
+ * `InjectionToken`, and will throw an error if not called from such a context.
```
most people will not use InjectionToken to interact with this. Maybe expand this to say "one defined via @Injectable factory or InjectionToken" and link both the the respective api pages.

mhevery replied:
+1 for documentation with example

alxhub replied:
`@Injectable` does not yet have `factory` (you could use `inject` in `useFactory` but that's not a pattern that we've decided to teach yet).

I've added an example that shows how to use `inject` within an `InjectionToken` function.

I'll communicate this public API change to the docs team working on tree-shakeable providers as well, and they can document `inject` in their new material.

IgorMinar commented on packages/compiler-cli/integrationtest/bazel/injectable_def/app/src/dep.ts:
```diff
@@ -0,0 +1,41 @@
+/**
```
since you are proposing to make inject a public api, you should also make use of it in this test. currently all the testing is done indirectly by relying on the generating code containing "inject" call.

alxhub replied:
I've added it to the `token` test.

mary-poppins commented:
You can preview b583797 at https://pr22389-b583797.ngbuilds.io/.

mary-poppins commented:
You can preview 3c741c7 at https://pr22389-3c741c7.ngbuilds.io/.

mary-poppins commented:
You can preview b52f128 at https://pr22389-b52f128.ngbuilds.io/.

mary-poppins commented:
You can preview 4086cb3 at https://pr22389-4086cb3.ngbuilds.io/.

IgorMinar commented on packages/examples/core/di/ts/injector_spec.ts:
```diff
@@ -37,5 +55,25 @@ import {InjectionToken, Injector, ReflectiveInjector} from '@angular/core';
+      const injector = new MockRootScopeInjector(ReflectiveInjector.resolveAndCreate([Dep]));
+
+      // #docregion ShakeableInjectionToken
+      class Service {
```
please don't use generic variable/symbol names in examples because for an inexperienced person it's hard to distinguish between the variable/symbol names and keywords/apis.

For example, use "MyService" instead of "Service" and "MyDependency" and "myDependency" instead of "dep" and "Dep".

mary-poppins commented:
You can preview 07a6e86 at https://pr22389-07a6e86.ngbuilds.io/.

IgorMinar commented:
@alxhub thanks for making the requested changes.
